initrd/bin/kexec-seal-key

```diff
@@ -70,59 +70,65 @@ dd \
-	# We need to count the number of ENABLED slots for both LUKS1 and LUKS2
-	# create regex pattern for both LUKS1 and LUKS2
-	regex="Slot [0-9]*: ENABLED"
-	regex+="\|"
-	regex+="[0-9]*: luks2"
-	slots_used=$(cryptsetup luksDump "$dev" | grep -c "$regex" || die "Unable to get number of slots used on $dev")
-
-	DEBUG "Number of slots used on $dev LUKS header: $slots_used"
-	# If slot1 is the only one used, warn and die with proper messages
-	if [ "$slots_used" -eq 1 ]; then
-		# Check if slot 1 is the only one existing
-		if [ "$(cryptsetup luksDump "$dev" | grep -c "Slot 1: ENABLED")" -eq 1 ] || [ "$(cryptsetup luksDump "$dev" | grep -c "1: luks2")" -eq 1 ]; then
-			warn "Slot 1 is the only one existing on $dev LUKS header. Heads cannot use it to store TPM sealed LUKS Disk Unlock Key"
-			warn "Slot 1 should not be the only slot existing on $dev LUKS header. Slot 0 should be used to store LUKS Disk Recovery Key/passphrase"
-			die "You can safely fix this before continuing through Heads recovery shell: cryptsetup luksAddKey $dev"
-		fi
-	else
-		DEBUG "Slot 1 is not the only existing slot on $dev LUKS header."
-		DEBUG "$dev LUKS header's slot 1 will store LUKS Disk Unlock Key that TPM will seal/unseal with LUKS TPM Disk Unlock Key passphrase"
+	DEBUG "Testing $RECOVERY_KEY keyfile created from provided passphrase against $dev"
+	if ! cryptsetup luksOpen --test-passphrase --key-file "$RECOVERY_KEY" $dev; then
+		die "$dev: Unable to unlock LUKS device with privided passphrase. Exiting..."
 	fi
-done
 
-# Remove all the old keys from slot 1
-for dev in $(cat "$KEY_DEVICES" | cut -d\  -f1); do
-	echo "++++++ $dev: Removing old LUKS TPM Disk Unlock Key in LUKS slot 1"
-	cryptsetup luksKillSlot \
-		--key-file "$RECOVERY_KEY" \
-		$dev 1 ||
-		warn "$dev: removal of LUKS TPM Disk Unlock Key in LUKS slot 1 failed: might not exist. Continuing"
-
-	echo "++++++ $dev: Adding LUKS TPM Disk Unlock Key to LUKS slot 1"
-	cryptsetup luksAddKey \
-		--key-file "$RECOVERY_KEY" \
-		--key-slot 1 \
-		$dev "$KEY_FILE" ||
-		die "$dev: Unable to add LUKS TPM Disk Unlock Key to LUKS slot 1"
+	# We will use the following regex to find the slots that are enabled
+	regex="Slot ([0-9]+): ENABLED|\b([0-9]+): luks2"
+
+	# good_slot will be the slot number where the passphrase was tested against as valid. We will keep that slot
+	good_slot=-1
+
+	# test each possible existing key slot of dev against keyfile $RECOVERY_KEY
+	for slot in $(cryptsetup luksDump "$dev" | grep -E "$regex" | sed -r 's/Slot ([0-9]+): ENABLED|\b([0-9]+): luks2/\1\2/'); do
+		if [ "$good_slot" -eq "-1" ]; then
+			if cryptsetup luksOpen --test-passphrase --key-file "$RECOVERY_KEY" $dev $slot; then
+				good_slot="$slot"
+				break;
+			fi
+		fi
+	done
+
+	# if we found a good slot, we wipe all the other slots on current $dev
+	for slot in $(cryptsetup luksDump "$dev" | grep -E "$regex" | sed -r 's/Slot ([0-9]+): ENABLED|\b([0-9]+): luks2/\1\2/'); do
```

Comment:
> Why are we wiping keyslots at all

Well. As of today, Heads is the responsible of adding/removing keyslots and dealing with the keyslots. Only live tool I'm aware of is QubesOS "Change Disk Password" which does the right thing and resuses the same keyslot. As for cleaning up mess, it's important to make sure we just not postpone the exhaustion of keyslot forever. I'm not aware of all cryptsetup use cases out there. but under Heads context. only two keyslots should be used: The DRK (which cryptsetup version bump doesn't permit reusing which is why this logic is changing) where prior user's complaints were about having used external tools, without specifying keyslot, which moved DRK into slot 1 and in this really corner use case, the luksWipe was wiping the only keyslot used, which was keyslot 1. So under master now, there is a check to make sure no luksWipe would happen if we have only one keyslot to make sure Heads is not ever the culprit into wiping keyslots. Under cryptsetup 2.6.1 version bump, as said, we now know that reencryption will rotate keyslot and put DRK in next avail slot. So the new logic needs to not take for granted which keyslot is the DRK and test the keyslot against user provided DRK passphrase. As to now answer your question about why wiping keyslots, as long as a keyslot is determined to be the DRK, all other slots are to be rendered invalid anyway since DRK has changed in case of reencryption anyway. For this specific case where we are adding a DUK, again leaving the previous ones not being the DRK would be bogus as well, since the LUKS header changed, and therefore any other slot is also deprecated. The only cases not taken into consideration here is if for whatever reason, another DRK is used, which again AFAIK should not happen. We can improve this upon other use cases in other PRs, but consuming all keyslots and causing issues later on is not an option

Reply:
> Potentially problematic regex (sed) in opening statement, see earlier review comment

Yeah. I think I will use luksDump output and use "Version:" and then seperate the logic, which would not deprecate luks1 (we don't know what people use under Heads, which is privacy enforcing but also problematic). A quick search showed that debian9 and ubuntu 18.04 were deployed in luksv1 and might have been upgraded in place. Same for QubesOS 4.0 upgraded in place to Q4.2.1. We don't know. So I will rework that part to not drop luks1 and deploy an older debian locally under qemu and test codepath against newer and older luks containers....

Reply:
I'll reply to your [first reply here](https://github.com/linuxboot/heads/pull/1541/files#r1580151709):
I'm still not grasping why it goes through all occupied slots to wipe them...from your reply it seems these are the relevant parts:
> all other slots are to be rendered invalid anyway since DRK has changed in case of reencryption anyway

But, from the cryptsetup 2.6.1 man page about reencryption with `--key-slot` parameter:
```--key-slot, -S <0-N>
           For LUKS operations that add key material, this option allows you to specify which key slot is selected for the new key.
           For reencryption mode it selects specific keyslot (and passphrase) that can be used to unlock new volume key. If used all other keyslots get removed after reencryption operation is finished.
```
And AFAICT all re-encryption calls in Heads use the `--key-slot` parameter, so all other key slots should already have been wiped in that case anyway (by `cryptsetup reencrypt --keyslot [...]`).

The only other reasoning seems to be (please correct me if I'm wrong):
> it's important to make sure we just not postpone the exhaustion of keyslot forever

I wrote in my original review comment here (about `cryptsetup reencrypt`):
> after putting the new volume key in the new slot, cryptsetup will wipe the old slot

I have tested this (as I also wrote)...so I don't understand how such a "keyslot exhaustion" could happen. The only way would be if the user does `cryptsetup luksAddKey` 30 times, which is an extreme edge case.

Also, what if the DRK is in key slot 8?

----------------------------
IMO it would be better if Heads saves the output of the evaluated statement on line 86 (`cryptsetup luksDump "$dev" | grep -E "$regex" | sed -r 's/Slot ([0-9]+): ENABLED|\b([0-9]+): luks2/\1\2/'`) in an array (this is disregarding the regex discussion), then iterates over that, checking where the DRK is and then wipes key slot 1 and/or 8 (both if DRK is neither in 1 nor 8, otherwise the one where it isn't); then it puts the TPM DUK in a key slot other than the DRK keyslot (either 1 or 8...default can be 8 and 1 could be used in the case that the DRK is in slot 8). I don't think Heads needs to bother with wiping other key slots.

Reply:
Houla. We have a big misunderstanding here. If you reencrypt. The data changes, right, since the key change, not just its passphrase. This is the reasosn why we push for OEMs to push for Re-Ownership with reencryption. So any other keyslot being used here after reencryption is obsolete. And reencryption rotates keyslot to the next available one.

So accumulating keyslots usage without wiping them is not an option.

Now, luksAddkey adds keyslots until exhaustion. We need to wipe them.

After that, Heads has no idea of what else is consuming keyslot other then the one that can be unllocked with DRK,
The logic I try to push forward is having DRK keyslot being dynamic instead of fixated, and having DUK into last keyslot availabe per LUKS version, wiping any other one that was not DRK.

As for if that dUK keyslot was the last one per LUKS version, then Heads would bail, asking the user to do manual changes. There is no reason today that key should consume last keyslot per LUKS version.


---
What would happen if keyslots get exhausted, please? How would the user deal with that if noone does @UndeadDevel ? I was going to implement changes but those comments are just making me want to take another step back.

Goal
- have reencryption do keyslot +1 and not care whenre that land
- test disk agaisnt provided user DRK passphrase, wipe all other slots

That's it.
This way, addink a DUK (that's the codepath we're talking here) don't lead to keyslot exhaustion. Otherwise how would you deal with possible keyslot exhaustion.

I would love to adapt code to fix this issue once and for all beginning of next week and move forward.

Reply:
Yes, we do seem to be talking past one another; maybe I'm not being clear enough, but I've spent already quite a bit of time here, so I guess I'll just wait for the next iteration of your code rather than repeat the same again.

Reply:
I'll try once more to clarify @tlaurion:

Yes, we need to wipe the old TPM DUK slot outside of re-encryption events, but there is no need to wipe all slots, just the previous one; if slots 1 and 8 are designated for this purpose then I've laid out a possible logic above already, if not, then it depends on the new algorithm. 
For the re-encryption case, cryptsetup already wipes the old slots on re-encrypt automatically according to the cryptsetup 2.6.1 man page, so Heads code doesn't need to bother about it.

Reply:
Testing unfortunately proved otherwise. Reencrypting reuses the same slot and needs that slot to be specified, where passphrase change rotates the slot. 

So current PR code does both.
- test passphrase to see if unlock is possible by simply testing passphrase against all luks containers of same drive (BTRFS qubes 4.2 setup used here). If not, proposes user to change this with options->change luks passphrase for containers not matching for which test failed
- For reencryption, test individual keyslots to find which one is hosting the DRK.
- For passphrase change, let cryptsetup does its thing and refactoring of code to not consider any DRK slot as ebing static
- When setting DUK, cover old codebase implementation for DUK being in either slot 1, or by LUKSv1 new code: 7 LUKSv2:31, and propose propose to wipe any other key slot not being DRK by individual testing.

Commit logs detail the changes, this is just to remind of implementation choice. As said there, if we were to let DUK add keys in additional key slots forever, eventually changing passphrase would fail and therefore we would need to do cleanup anyway prior of LUKS operation. So as of now, Heads is considered the sole LUKS key slot modififier, where if GPG LUKS key (Purism implementation) is used (@JonathonHall-Purism) then if user uses DUK+Purism implementation, Heads will warn user that a key slot outside of DRK and Heads LUKSv1:7/LUKSv2:31 was found and ask user to confirm wipe of it.

@UndeadDevel @JonathonHall-Purism that's the best I came up with. Something better to suggest?

Reply:
@UndeadDevel @JonathonHall-Purism refactored.